rd-net/Lifetimes/Util/Memory.cs

```diff
@@ -52,14 +52,59 @@ public static unsafe void CopyMemory(byte* src, byte* dest, int len)
       }
     }
 
+    [MethodImpl(MethodImplAdvancedOptions.AggressiveInlining)]
     public static void Barrier()
     {
 #if (NET35)
   Thread.MemoryBarrier();
 #else
   Interlocked.MemoryBarrier();
 #endif
-  
+    }
+    
+    [MethodImpl(MethodImplAdvancedOptions.AggressiveInlining)]
+    public static T VolatileRead<T>(ref T location) where T : class
+    {
+#if (NET35)
+      Thread.MemoryBarrier();
+      return location;
```

Comment:
First: mind `Thread.VolatileRead` static method (which isn't that performant on .NET Framework, but is basically in line with your intentions).

Second: I think that this implementation is incorrect. Inside of a volatile read, you're supposed to _first_ read the value, and _then_ place a memory barrier.

In .NET Framework, the method is implemented as explained by J. Albahari [in his article](http://www.albahari.com/threading/part4.aspx) (seek for `VolatileRead`).

So, I suggest to just fall back to `Thread.VolatileRead`/`VolatileWrite`, if possible (you'll have to degeneralize the code, since there're no generic versions of those, but I believe that's okay).

(your implementation of `VolatileWrite` looks correct, though)

Reply:
Hm, You're right. First of all, I tried to avoid caching the variable by jit. Because I had a sad experience with deadlocks because of it. But looks like jit works as expected in this place.